packages/protocol/contracts/psp/README.md

```diff
@@ -0,0 +1,23 @@
+## PSP Notes
+
+Currently, most DeFi products depend on a live data point provided by a passive oracle service, and keepers to trigger actions automatically based on the data provided by the oracle, e.g., liquidate a user that has shorted an asset that has increased in price. PSP allows Airnode to act as an oracle and a keeper rolled into one, and provide services that are not available with the oracle + keeper combos that are currently available. From the API integration point-of-view, it imititates an on-chain webhook API, yet does not require the underlying API to be nothing more than a regular REST API.
+
+Even though PSP will enable oracle usage patterns that were not available before, it will also provide a much more permissionless and thus decentralized way of setting up traditional asset price data feeds. In short, one can subscribe to a set of Airnodes to keep individual price feeds up-to-date on a PspBeaconServer contract and update the aggregation when necessary. Although data feeds currently exist, they are not set up through a protocol, but rather through mails being exchanged and meatspace agreements being signed, which is extremely inefficient. Protocolizing setting up first-party oracle-based data feeds across all chains will be a game-changer.
+
+The main difficulty in providing a subscription service is that an API/Airnode can serve a limited number of them at a time (compared to this, RRP is much more scalable). Then, the resource that we must allow the Airnode to manage are these subscription "slots". This slot allocation needs to happen on-chain, programmatically. To achieve this, we implement "allocators", contracts similar to "authorizers" that implement arbitrary business logic that the Airnode operator can choose to opt-in to by adding them in their config file. Note that a requester both needs to have a slot allocated by an allocator, and be authorized by an authorizer to receive service.
+
+The authorizer parameters will be filled as such:
+- `requestId`: `subscriptionId`
+- `airnode`: `airnode`
+- `endpointId`: `endpointId`
+- `sponsor`: `sponsor`
+- `requester`: `fulfillAddress`
+Alternatively, we can refer to the subscriber as the requester (instead of the sponsor), and implement an additional sponsorship layer on top.
+
+For each chain, the config file should specify a list of allocators with the following fields:
+- Allocator chain + provider(s)
+- Allocator address
+- Maximum number of slots
```

Comment:
So this is global for the allocator? Shouldn't this be in the contract?

Reply:
Actually, do I understand this field correctly? This is the number that will be used as `limit` when calling `getActiveSubscriptions`.

I mean the limit could be optional in the contract (default value -1) and settable through admin CLI by the API provider? But this is a bit unfortunate that you have to make TX for this... Also, a big pain is to keep this in sync with what's in config json.

Can't you keep this `limit` number in sync in the contract instead? (update it when the maximum changes). Some allocator updates will be more costly then others, but I don't think it's a big deal.

Also, I wonder how will the allocator interface look like. Will it have only the `getActiveSubscriptions`? I guess that is all the Airnode needs... As mentioned above, I'd like if the `limit` (maximum number of slots of sponsor) was managed by the contract. I'd then imagine the following interface:

```solidity
interface IAllocator {
    function getTotalNumberOfActiveSlots() external view returns (uint256);

    function getMaxSlotNumberOfAirnode(address airnode) external view returns (uint256);

    function getNumberOfActiveSlotsOfAirnode(address airnode) external view returns (uint256);

    function getActiveSlots(address airnode)
        external
        view
        returns (bytes32[] memory subscriptionIds, address[] memory sponsors);

    // in case the number of active slots is large - pagination is required???
    // the node can decide if this is necessary based on "getMaxSlotNumberOfAirnode"
    function getActiveSlotsPage(address airnode, uint256 from, uint256 to)
        external
        view
        returns (bytes32[] memory subscriptionIds, address[] memory sponsors);
}
```